src/main/java/fi/hsl/transitdata/stopestimates/PubtransStopEstimatesFactory.java

```diff
@@ -0,0 +1,138 @@
+                                roiMessage.getCommon(),
+                                roiMessage.getTripInfo()
+                        ));
+            }
+            else {
+                return Optional.empty();
+            }
+        }
+        catch (Exception e) {
+            log.error("Failed to parse PubtransData from schema " + schema.toString(), e);
+            return Optional.empty();
+        }
+    }
+
+    private static boolean validate(final PubtransData pubtransData) {
+        return validateCommon(pubtransData.common) && validateTripInfo(pubtransData.tripInfo);
+    }
+
+    private static boolean validateCommon(PubtransTableProtos.Common common) {
+        if (common == null) {
+            log.error("No Common, discarding message");
+            return false;
+        }
+        if (!common.hasIsTargetedAtJourneyPatternPointGid()) {
+            log.error("No JourneyPatternPointGid, message discarded");
+            return false;
+        }
+        if (!common.hasTargetUtcDateTimeMs()) {
+            log.error("No TargetDatetime, message discarded");
+            return false;
+        }
+        if (common.getType() == 0) {
+            log.info("Event is for a via point, message discarded");
+            return false;
+        }
+        return true;
+    }
+
+    private static boolean validateTripInfo(PubtransTableProtos.DOITripInfo tripInfo) {
+        if (tripInfo == null) {
+            log.error("No tripInfo, discarding message");
+            return false;
+        }
+
+        if (!tripInfo.hasRouteId()) {
+            log.error("TripInfo has no RouteId, discarding message");
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Throws RuntimeException if data is not valid.
+     */
+    private static InternalMessages.StopEstimate toStopEstimate(final PubtransData pubtransData) throws Exception {
+        InternalMessages.StopEstimate.Builder builder = InternalMessages.StopEstimate.newBuilder();
+        builder.setSchemaVersion(builder.getSchemaVersion());
+
+        InternalMessages.TripInfo.Builder tripBuilder = InternalMessages.TripInfo.newBuilder();
+        tripBuilder.setTripId(Long.toString(pubtransData.tripInfo.getDvjId()));
+        tripBuilder.setOperatingDay(pubtransData.tripInfo.getOperatingDay());
+        tripBuilder.setRouteId(pubtransData.tripInfo.getRouteId());
+        tripBuilder.setDirectionId(pubtransData.tripInfo.getDirectionId());//Jore format
+        tripBuilder.setStartTime(pubtransData.tripInfo.getStartTime());
+        builder.setTripInfo(tripBuilder.build());
+        builder.setStopId(pubtransData.tripInfo.getStopId());
+        builder.setStopSequence(pubtransData.common.getJourneyPatternSequenceNumber());
+
+        InternalMessages.StopEstimate.Status scheduledStatus = (pubtransData.common.getState() == 3L) ?
+                InternalMessages.StopEstimate.Status.SKIPPED :
```

Comment:
Remember to merge the `InternalMessages.StopEstimate.Status.SCHEDULED` temporary fix version.

Reply:
We can check this together.